Common/src/Common/Service/Data/Search/SearchTypeManager.php

```diff
@@ -23,4 +23,28 @@ class SearchTypeManager extends AbstractPluginManager
      * @var bool
      */
     protected $autoAddInvokableClass = false;
+
+    private array $config;
+
+    public function __construct(ContainerInterface $container, $config)
+    {
+        $this->config = $config;
+        parent::__construct($container, $config);
+    }
+
+    /**
+     * @todo this isn't great, but provides for backward compatibility with old code as we move to Laminas 3
+     *
+     * the old code (mis)used a debug method within the Laminas service manager to provide this info,
+     * for now we're replicating that behaviour here
+     */
+    public function getRegisteredServices(): array
+    {
+        $factories = $this->config['factories'] ?? [];
+        $invokables = $this->config['invokables'] ?? [];
+
+        return array_keys(
+            array_merge($factories, $invokables)
+        );
```

Comment:
We don't need to grab from config do we?

```suggestion
        return array_keys($this->factories);
```

Reply:
It's because it runs array_keys to get the info that it needs, so it can't get it from $this->factories. It could however get it from $this->aliases if that's what we wanted to do.

This approach exactly replicated what we had before and removed the chance of future service manager changes messing things up again (in the way which has just happened), so that's why I went down that route. But yes we could remove a few lines of code if we used $this->aliases instead, if that's the preference I can make the changes.

```php
        $factories = $this->config['factories'] ?? [];
        $invokables = $this->config['invokables'] ?? [];

        print_r(array_merge($factories, $invokables));
        print_r($this->aliases);
        print_r($this->factories);
```

Gives us:

```php
Array
(
    [licence] => Common\Data\Object\Search\Licence
    [application] => Common\Data\Object\Search\Application
    [case] => Common\Data\Object\Search\Cases
    [psv_disc] => Common\Data\Object\Search\PsvDisc
    [vehicle] => Common\Data\Object\Search\Vehicle
    [address] => Common\Data\Object\Search\Address
    [bus_reg] => Common\Data\Object\Search\BusReg
    [people] => Common\Data\Object\Search\People
    [user] => Common\Data\Object\Search\User
    [publication] => Common\Data\Object\Search\Publication
    [irfo] => Common\Data\Object\Search\IrfoOrganisation
)
Array
(
    [licence] => Common\Data\Object\Search\Licence
    [application] => Common\Data\Object\Search\Application
    [case] => Common\Data\Object\Search\Cases
    [psv_disc] => Common\Data\Object\Search\PsvDisc
    [vehicle] => Common\Data\Object\Search\Vehicle
    [address] => Common\Data\Object\Search\Address
    [bus_reg] => Common\Data\Object\Search\BusReg
    [people] => Common\Data\Object\Search\People
    [user] => Common\Data\Object\Search\User
    [publication] => Common\Data\Object\Search\Publication
    [irfo] => Common\Data\Object\Search\IrfoOrganisation
)
Array
(
    [Common\Data\Object\Search\Licence] => Laminas\ServiceManager\Factory\InvokableFactory Object
        (
        )
    [Common\Data\Object\Search\Application] => Laminas\ServiceManager\Factory\InvokableFactory
    [Common\Data\Object\Search\Cases] => Laminas\ServiceManager\Factory\InvokableFactory
    [Common\Data\Object\Search\PsvDisc] => Laminas\ServiceManager\Factory\InvokableFactory
    [Common\Data\Object\Search\Vehicle] => Laminas\ServiceManager\Factory\InvokableFactory
    [Common\Data\Object\Search\Address] => Laminas\ServiceManager\Factory\InvokableFactory
    [Common\Data\Object\Search\BusReg] => Laminas\ServiceManager\Factory\InvokableFactory
    [Common\Data\Object\Search\People] => Laminas\ServiceManager\Factory\InvokableFactory
    [Common\Data\Object\Search\User] => Laminas\ServiceManager\Factory\InvokableFactory
    [Common\Data\Object\Search\Publication] => Laminas\ServiceManager\Factory\InvokableFactory
    [Common\Data\Object\Search\IrfoOrganisation] => Laminas\ServiceManager\Factory\InvokableFactory
)
```

Reply:
Aliases eventually map back to `factories`, have you tried with just returning `array_keys($this->factories);`?

The code passes the keys to the service manager to retrieve so should resolve via. factories too.

Reply:
We can't do it with $this->factories as factories uses the class name (see example above).

The aliases uses a string key, which is what the search code is expecting. The search code uses the string for both the index names and for picking out the correct navigation key.

Below is array_keys for aliases vs factories

```php
Array
(
    [0] => licence
    [1] => application
    [2] => case
    [3] => psv_disc
    [4] => vehicle
    [5] => address
    [6] => bus_reg
    [7] => people
    [8] => user
    [9] => publication
    [10] => irfo
)
Array
(
    [0] => Common\Data\Object\Search\Licence
    [1] => Common\Data\Object\Search\Application
    [2] => Common\Data\Object\Search\Cases
    [3] => Common\Data\Object\Search\PsvDisc
    [4] => Common\Data\Object\Search\Vehicle
    [5] => Common\Data\Object\Search\Address
    [6] => Common\Data\Object\Search\BusReg
    [7] => Common\Data\Object\Search\People
    [8] => Common\Data\Object\Search\User
    [9] => Common\Data\Object\Search\Publication
    [10] => Common\Data\Object\Search\IrfoOrganisation
)
```

My thinking was by keeping this reliant on 5-10 lines of our own code, we don't run the risk of future ServiceManager changes breaking things again. But really not bothered either way, just need a decision one way or the other and will go with that.

Reply:
I see what you mean, but the only place this is used is here:

```php
$services = $this->getSearchTypeManager()->getRegisteredServices();

// ...

foreach ($services as $searchIndexName) {
    $indexes[] = $this->getSearchTypeManager()->get($searchIndexName);
}
```

The result from `getRegisteredServices` is just being fed back into the container, so it should work with FQCN too.

`invokables` behind the scenes are just populating `$this->factories` and `$this->aliases` in the service manager.

Using the `$factories` also future proofs it in my head. We can remove the aliases (/invokables) and use the FQCN without refactoring anything here. We can also swap out classes to use their own factory (if we ever needed to) a little easier.

Reply:
See what you mean now - crossed wires. I've altered it to use array_keys($factories), have removed the @todo as it no longer seems pressing to get rid of it...